Improve readme (#1114)
eclipse-thingweb/node-wot: README.md

    -   [Protocol Support](#protocol-support)
    -   [MediaType Support](#mediatype-support)
-   [No time for explanations - show me a running example!](#no-time-for-explanations---show-me-a-running-example)
    -   [Using Node.js](#using-nodejs)
    -   [Using Docker](#using-docker)
    -   [Using a browser](#using-a-browser)
-   [Online Things](#online-things)
-   [Documentation](#documentation)
    -   [The API](#the-api)
    -   [TD Tooling](#td-tooling)
    -   [Logging](#logging)
    -   [Install new/different versions of Node.js](#install-newdifferent-versions-of-nodejs)
-   [Contributing](#contributing)
-   [License](#license)

## Installation

The framework can be used in two ways: as a library or as a CLI tool. In this section we will explain how to install the framework in both ways.

### As a library

The framework is composed by different packages that users can use as they please. The core package is `@node-wot/core` and it is the only mandatory package to install. The other packages are bindings that allow the framework to communicate with different protocols and optionally we offer a set of tools in the `@node-wot/td-tools` package.

#### Node.js

> > **Warning**: We no longer actively support Node.js version 16 and lower.

Reviewer: Maybe we could format this using GitHub's relatively recent [alert syntax](https://github.com/orgs/community/discussions/16925)? :)

```suggestion
> [!WARNING]
> We no longer actively support Node.js version 16 and lower.
```

-   Microsoft Edge 15 and later
-   Firefox 54 and later
-   Chrome 58 and later
-   Safari 10 and later

Using a browser with only ES5 support (eg. IE 11) might be possible if you add polyfills. If you want to use node-wot as a library in your browser application, you can install the `@node-wot/browser-bundle` as following:

```
npm i @node-wot/browser-bundle --save
```

you can find more installation options in the specific [package README](./packages/browser-bundle/README.md).

### As a CLI tool

You can alternatively use node-wot via its command line interface (CLI). Please visit the [CLI tool's Readme](<[url](https://github.com/eclipse-thingweb/node-wot/tree/master/packages/cli)>) to find out more.

#### As a docker image

Another option is to use node-wot inside a docker image. Make sure you are under Linux or under WSL if you are running on Windows.

Clone the repository:

```
git clone https://github.com/eclipse-thingweb/node-wot
```

Go into the repository:

```
cd node-wot
```

Build the Docker image named `wot-servient` from the `Dockerfile`:

```
npm run build:docker
```

Run the wot-servient as a container:

```
docker run --rm wot-servient -h
```

## Examples

> With node-wot you can create server side Things, in WoT gergo we call this operation "expose a thing" or you can create client side Things, in WoT gergo we call this operation "consume a thing". In the following sections we will show how to create a simple counter Thing and how to consume it. Supposing you installed and configured node-wot as a library, you can create and expose counter Thing as follows:

Author: ```suggestion
With node-wot you can create server-side Things, in WoT gergo we call this operation "expose a thing" or you can create client-side Things, in WoT gergo we call this operation "consume a thing". An expose thing allows you to bring your device or services to the Web with just a few lines of code. On the other hand, with a consume thing you have a fixed interface to interact with devices on different protocols/frameworks.  In the following section, we will show how to create a simple counter Thing and how to consume it. Supposing you installed and configured node-wot as a library, you can create and expose counter Thing as follows:
```

Suggested by @hyperloris .

Reviewer: Looks good to me! I spotted a few nits, though, that I tried to incorporate in this updated suggestion:

```suggestion
With node-wot you can create server-side Things, in WoT jargon we call this operation "expose a thing" or you can create client-side Things, in WoT gergo we call this operation "consume a thing". An exposed Thing allows you to bring your device or services to the Web with just a few lines of code. On the other hand, with a consumed Thing you have a fixed interface to interact with devices using different protocols/frameworks. In the following section, we will show how to create a simple counter Thing and how to consume it. Assuming you have installed and configured node-wot as a library, you can create and expose counter Thing as follows:
```

> <img title="ThingWeb" alt="Thingweb logo" src="https://raw.githubusercontent.com/eclipse-thingweb/website/master/misc/thingweb_logo.svg" width=300px>

Reviewer 2: ```suggestion
<img title="ThingWeb" alt="Thingweb logo" src="https://raw.githubusercontent.com/eclipse-thingweb/website/master/misc/thingweb_logo.svg" width="300px">
```

Since we did it elsewhere...

> ## Node-wot

Reviewer 2: ```suggestion
## node-wot
```

lower-case?

Author: checked also playground is upper case. committing.

> A fast and extensible framework to connect any device with your application

[![Default CI Pipeline](https://github.com/eclipse-thingweb/node-wot/actions/workflows/ci.yaml/badge.svg)](https://github.com/eclipse-thingweb/node-wot/actions/workflows/ci.yaml)
[<img alt="npm" src="https://img.shields.io/npm/dw/@node-wot/td-tools">](https://npm-stat.com/charts.html?package=%40node-wot%2Ftd-tools)
[![codecov](https://codecov.io/gh/eclipse-thingweb/node-wot/branch/master/graph/badge.svg)](https://codecov.io/gh/eclipse-thingweb/node-wot)
[![Telegram Group](https://img.shields.io/endpoint?color=neon&url=https%3A%2F%2Ftg.sumanjay.workers.dev%2Fnodewot)](https://t.me/nodewot)
[![Discord](https://img.shields.io/badge/Discord-7289DA?logo=discord&logoColor=white&label=node-wot)](https://discord.gg/JXY2Jzefz3)

> The Eclipse Thingweb node-wot is a framework for implementing [Web of Things](https://www.w3.org/WoT/) servers and clients in Node.js. It is written from the ground up with Typescript with the goal of providing a fast and extensible framework for IoT applications. Node-wot wants to give developers the ability to create complex business logic without worring about protocol and low level details levereging to a standard metadata format, the [Thing Description (TD)](https://www.w3.org/TR/wot-thing-description11/). Thanks to the TD abstraction developers can find a set of satellite tools to create their applications in a fast and easy way.

Reviewer 3: ```suggestion
The Eclipse Thingweb node-wot is a framework for implementing [Web of Things](https://www.w3.org/WoT/) servers and clients in Node.js. It is written from the ground up with Typescript with the goal of providing a fast and extensible framework for IoT applications. Node-wot wants to give developers the ability to create complex business logic without worrying about protocol and low-level details leveraging a standard metadata format, the [Thing Description (TD)](https://www.w3.org/TR/wot-thing-description11/). Thanks to the TD abstraction, developers can find a set of satellite tools to create their applications in a fast and easy way.
```

Author: I think it should be like this.

```suggestion
The Eclipse Thingweb node-wot is a framework for implementing [Web of Things](https://www.w3.org/WoT/) servers and clients in Node.js. It is written from the ground up with Typescript with the goal of providing a fast and extensible framework for IoT applications. Node-wot wants to give developers the ability to create complex business logic without worrying about protocol and low-level details leveraging on a standard metadata format, the [Thing Description (TD)](https://www.w3.org/TR/wot-thing-description11/). Thanks to the TD abstraction developers can find a set of satellite tools to create their applications in a fast and easy way.
```

## Web of Things principles in a Nutshell

> The Web of Things (WoT) tackles IoT fragmentation by extending standardized web technologies. It simplifies IoT application development, promoting flexibility and interoperability. WoT preserves existing IoT standards, ensuring reuse, and provides an adaptable, abstract architecture based on real-world use cases across domains. In essence, WoT paves the way for seamless IoT integration by defining a metadata model capable of describing Things and Services and how to interact with them. This metadata model is called Thing Description (TD) and it is a JSON-LD document that describes the Thing and its capabilities, including its services (APIs), its network interactions, and security requirements. The TD is the cornerstone of the Web of Things architecture and it is the main abstraction that node-wot uses to implement its functionalities. Every Thing has the following capabilities or "affordances":

Reviewer 3: ```suggestion
The Web of Things (WoT) tackles IoT fragmentation by extending standardized web technologies. It simplifies IoT application development, promoting flexibility and interoperability. WoT preserves existing IoT standards, ensuring reuse, and provides an adaptable, abstract architecture based on real-world use cases across domains. In essence, WoT paves the way for seamless IoT integration by defining an information model capable of describing Things and Services and how to interact with them. This information model is called the Thing Description (TD) and it is a JSON-LD document that describes the Thing and its capabilities, including its network services (APIs), its network interactions, and security requirements. The TD is the cornerstone of the Web of Things architecture and it is the main abstraction that node-wot uses to implement its functionalities. Every Thing has the following capabilities or "affordances":
```

Reviewer 3: metamodel should be avoided since it can be misunderstood to be something way too abstract

#### Node.js

> [!WARNING]
> We no longer actively support Node.js version 16 and lower.

-   [Node.js](https://nodejs.org/) version 18+
-   [npm](https://www.npmjs.com/) version 9+

Platforms specific prerequisites:

-   Linux: Meet the [node-gyp](https://github.com/nodejs/node-gyp#installation) requirements:
    -   Python v3.6, v3.7, or v3.8
    -   make
    -   A proper C/C++ compiler toolchain, like GCC
-   Windows: Install the Windows build tools through a CMD shell as administrator:
    -   `npm install -g --production windows-build-tools`
-   Mac OS: Meet the [node-gyp](https://github.com/nodejs/node-gyp#installation) requirements:
    -   `xcode-select --install`

If you want to use node-wot as a library in your Node.js application, you can use npm to install the node-wot packages that you need. To do so, `cd` inside your application folder, and run:

```
npm i @node-wot/core @node-wot/binding-http --save
```

#### Browser

To use node-wot as a browser-side JavaScript Library, the browser needs to support ECMAScript 2015.
Supported browsers include:

-   Microsoft Edge 15 and later
-   Firefox 54 and later
-   Chrome 58 and later

> -   Safari 10 and later

Reviewer 2: Do we actually need the list of browsers and version? I would stick to the first sentence "To use node-wot as a browser-side JavaScript Library, the browser needs to support ECMAScript 2015." only

Author: Considering that we didn't actually test node-wot in all these browsers and we don't have compliance tests I agree to remove it.

Author: I can't do it easly now, maybe in a followup quick PR?

Reviewer 2: created https://github.com/eclipse-thingweb/node-wot/issues/1115

> Using a browser with only ES5 support (eg. IE 11) might be possible if you add polyfills. If you want to use node-wot as a library in your browser application, you can install the `@node-wot/browser-bundle` as following:

Reviewer 2: abbreviation is "e.g." or "e.g.," I think

Author: ```suggestion
Using a browser with only ES5 support (e.g., IE 11) might be possible if you add polyfills. If you want to use node-wot as a library in your browser application, you can install the `@node-wot/browser-bundle` as following:
```

```
npm i @node-wot/browser-bundle --save
```

you can find more installation options in the specific [package README](./packages/browser-bundle/README.md).

### As a CLI tool

You can alternatively use node-wot via its command line interface (CLI). Please visit the [CLI tool's Readme](<[url](https://github.com/eclipse-thingweb/node-wot/tree/master/packages/cli)>) to find out more.

#### As a docker image

Another option is to use node-wot inside a docker image. Make sure you are under Linux or under WSL if you are running on Windows.

Clone the repository:

```
git clone https://github.com/eclipse-thingweb/node-wot
```

Go into the repository:

```
cd node-wot
```

Build the Docker image named `wot-servient` from the `Dockerfile`:

```
npm run build:docker
```

Run the wot-servient as a container:

```
docker run --rm wot-servient -h
```

## Examples

> With node-wot you can create server-side Things, in WoT jargon we call this operation "expose a thing" or you can create client-side Things, in WoT gergo we call this operation "consume a thing". An exposed Thing allows you to bring your device or services to the Web with just a few lines of code. On the other hand, with a consumed Thing, you have a fixed interface to interact with devices, potentially using different protocols/frameworks. In the following section, we will show how to create a simple counter Thing and how to consume it. Assuming you have installed and configured node-wot as a library, you can create and expose a counter Thing as follows:

Reviewer: Just noticed another nit (although the Italian version also sounds great 🙂)

```suggestion
With node-wot you can create server-side Things, in WoT jargon we call this operation "expose a thing" or you can create client-side Things, in WoT jargon we call this operation "consume a thing". An exposed Thing allows you to bring your device or services to the Web with just a few lines of code. On the other hand, with a consumed Thing, you have a fixed interface to interact with devices, potentially using different protocols/frameworks. In the following section, we will show how to create a simple counter Thing and how to consume it. Assuming you have installed and configured node-wot as a library, you can create and expose a counter Thing as follows:
```

Author: ahahha I thought it was Latin or something :D :D committing. Thanks, Jan!

> With node-wot you can create server-side Things, in WoT jargon we call this operation "expose a thing" or you can create client-side Things, in WoT jargon we call this operation "consume a thing". An exposed Thing allows you to bring your device or services to the Web with just a few lines of code. On the other hand, with a consumed Thing, you have a fixed interface to interact with devices, potentially using different protocols/frameworks. In the following section, we will show how to create a simple counter Thing and how to consume it. Assuming you have installed and configured node-wot as a library, you can create and expose a counter Thing as follows:

Reviewer 3: ```suggestion
With node-wot you can create server-side Things, in WoT jargon we call this operation "expose a Thing" or you can create client-side Things, in WoT jargon we call this operation "consume a Thing". An exposed Thing allows you to bring your device or services to the Web with just a few lines of code. On the other hand, with a consumed Thing, you have a fixed interface to interact with devices, potentially using different protocols/frameworks. In the following section, we will show how to create a simple counter Thing and how to consume it. Assuming you have installed and configured node-wot as a library, you can create and expose a counter Thing as follows:
```

```JavaScript
// Required steps to create a servient for creating a thing
const Servient = require('@node-wot/core').Servient;
const HttpServer = require('@node-wot/binding-http').HttpServer;

const servient = new Servient();
servient.addServer(new HttpServer());

servient.start().then( async (WoT) => {
    // Then from here on you can use the WoT object to produce the thing
    let count = 0;
    const exposingThing = await WoT.produce({
        title: "Counter",
        description: "A simple counter thing",
        properties: {
            count: {
                type: "integer",
                description: "current counter value",
                observable: true,
                readOnly: true
            }
        },
        actions: {
            increment: {
                description: "increment counter value",
            }
        }
    })
    exposingThing.setPropertyReadHandler("count", () => { return count; });
    exposingThing.setActionHandler("increment", () => { count++; exposingThing.emitPropertyChange("count"); });
    await exposingThing.expose();
    // now you can interact with the thing via http://localhost:8080/counter
});
```

> Now supposing you want to interact with the device you have to consume its thing description as follows:

Reviewer 3: ```suggestion
Now supposing you want to interact with the device, you have to consume its Thing Description as follows:
```

-   [CoAP](https://github.com/eclipse-thingweb/node-wot/blob/master/packages/binding-coap/README.md) :heavy_check_mark:
-   [CoAPS](https://github.com/eclipse-thingweb/node-wot/blob/master/packages/binding-coap/README.md) :heavy_check_mark:
-   [MQTT](https://github.com/eclipse-thingweb/node-wot/blob/master/packages/binding-mqtt/README.md) :heavy_check_mark:
-   ~~[Firestore//lastSupportedVersion](https://github.com/eclipse/thingweb.node-wot/blob/v0.8.6/packages/binding-firestore/README.md) :heavy_check_mark:~~
-   [Websocket](https://github.com/eclipse-thingweb/node-wot/tree/master/packages/binding-websockets) :heavy_plus_sign: (Server only)
-   [OPC-UA](https://github.com/eclipse-thingweb/node-wot/blob/master/packages/binding-opcua/README.md) :heavy_plus_sign: (Client only)
-   [NETCONF](https://github.com/eclipse-thingweb/node-wot/blob/master/packages/binding-netconf/README.md) :heavy_plus_sign: (Client only)
-   [Modbus](https://github.com/eclipse-thingweb/node-wot/tree/master/packages/binding-modbus) :heavy_plus_sign: (Client only)
-   [M-Bus](https://github.com/eclipse-thingweb/node-wot/tree/master/packages/binding-mbus) :heavy_plus_sign: (Client only)

> Note: More protocols can be easily added by implementing `ProtocolClient`, `ProtocolClientFactory`, and `ProtocolServer` interface.

Reviewer: Maybe we could also use alerts here?

```suggestion
> [!NOTE]
> More protocols can be easily added by implementing `ProtocolClient`, `ProtocolClientFactory`, and `ProtocolServer` interface.
```

Note: The bindings for [binding-fujitsu](https://github.com/eclipse-thingweb/node-wot/tree/v0.7.x/packages/binding-fujitsu) and [binding-oracle](https://github.com/eclipse-thingweb/node-wot/tree/v0.7.x/packages/binding-oracle) were removed after `v0.7.x` due to lack of maintainers.

### MediaType Support

-   JSON :heavy_check_mark:
-   Text (HTML, CSS, XML, SVG) :heavy_check_mark:
-   Base64 (PNG, JPEG, GIF) :heavy_check_mark:
-   Octet stream :heavy_check_mark:
-   CBOR :heavy_check_mark:
-   EXI :timer_clock:

> Don't you find your preferred MediaType? More codecs can be easily added by implementing `ContentCodec` interface. Read more in the [Documentation](#documentation) section.

Reviewer: ```suggestion
Can't find your preferred MediaType? More codecs can be easily added by implementing `ContentCodec` interface. Read more in the [Documentation](#documentation) section.
```

servient.addClientFactory(new HttpClientFactory(null));
const WoTHelpers = new Helpers(servient);

WoTHelpers.fetch("http://localhost:8080/counter").then(async (td) => {
    try {
        const WoT = await servient.start();
        // Then from here on you can consume the thing
        let thing = await WoT.consume(td);
        thing.observeProperty("count", async (output) => { console.log("count:", await data.value()); });
        for (let i = 0; i < 5; i++) {
            await thing.invokeAction("increment");
        }
    }
    catch (err) {
        console.error("Script error:", err);
    }
}).catch((err) => { console.error("Fetch error:", err); });
```

> If you execute both scripts you will see `count: ${count}` printed 5 times. We hosted a more complex version of this example at [http://plugfest.thingweb.io/examples/counter.html](http://plugfest.thingweb.io/examples/counter.html) and you can find the source code in the [counter example](./examples/browser) folder. You can also find more examples in the [examples folder](./examples/scripts) for Javascript and in the [examples folder](./packages/examples/) for Typescript. Finally, for your convenience we host a set of online Things that you can use to test your applications. You can find more information about them in the [Online Things](#online-things) section.

Reviewer 3: ```suggestion
If you execute both scripts you will see `count: ${count}` printed 5 times. We host a more complex version of this example at [http://plugfest.thingweb.io/examples/counter.html](http://plugfest.thingweb.io/examples/counter.html) and you can find the source code in the [counter example](./examples/browser) folder. You can also find more examples in the [examples folder](./examples/scripts) for JavaScript and in the [examples folder](./packages/examples/) for TypeScript. Finally, for your convenience, we host a set of online Things that you can use to test your applications. You can find more information about them in the [Online Things](#online-things) section.
```

-   Smart Coffee Machine: This is a simulation of a coffee machine that also has a [simple user interface](http://plugfest.thingweb.io/examples/smart-coffee-machine.html) that displays the values of properties.
    In addition to proving a real-life device example, it can be used for testing `uriVariables`. You can ask it to brew different coffees and monitor the available resource level.

## Documentation

> > **Warning** ⚒️ We are planning to extend this section and to provide a more detailed documentation. Stay tuned!

Reviewer: ```suggestion
> [!WARNING]
> ⚒️ We are planning to extend this section and to provide a more detailed documentation. Stay tuned!
```

To learn by examples, see `examples/scripts` to have a feeling of how to script a Thing or a Consumer.

### Adding a new codec

To add a new codec, you need to implement the `ContentCodec` interface. The interface is defined as follows:

```TypeScript
export interface ContentCodec {
    getMediaType(): string;
    bytesToValue(bytes: Buffer, schema: DataSchema, parameters?: {[key: string]: string}): any;
    valueToBytes(value: any, schema: DataSchema, parameters?: {[key: string]: string}): Buffer;
}
```

Finally you can add to your servient the new codec as follows:

```JavaScript
const ContentSerdes = require('@node-wot/core').ContentSerdes
const JsonCodec = require('@node-wot/core').JsonCodec

// e.g., assign built-in codec for *new* contentType

> let cs = ContentSerdes.get();

Reviewer: Tiny nit: 

```suggestion
const cs = ContentSerdes.get();
```

1. Run `npm dedupe` (see https://github.com/eclipse-thingweb/node-wot/pull/765#issuecomment-1133772886)

</details>

## License

> Dual-licensed under both

Reviewer 3: ```suggestion
Dual-licensed under:
```

I have removed both since if someone does not read the sentence later on, they can think of it as an AND relationship
